src/PathMappingRepository.php

```diff
@@ -0,0 +1,382 @@
+     */
+    private function resolveResource($path)
+    {
+        /*
+         * If the path exists in the store, return it directly
+         */
+        if ($this->store->exists($path)) {
+            $resolved = $this->createResource($this->store->get($path));
+            $resolved->attachTo($this, $path);
+
+            return $resolved;
+        }
+
+        /*
+         * Otherwise, we need to "resolve" it in two steps:
+         *      1.  find the resources from the store that are potential parent
+         *          of the path (we find them using regular expressions)
+         *      2.  for each of these potential parent, we try to find a real
+         *          file or directory on the filesystem and if we do find one,
+         *          we stop
+         */
+        $basePaths = array_reverse($this->store->keys());
+
+        foreach ($basePaths as $key => $basePath) {
+            if (!Path::isBasePath($basePath, $path)) {
+                continue;
+            }
+
+            $filesystemBasePath = rtrim($this->store->get($basePath), '/').'/';
+            $filesystemPath = substr_replace($path, $filesystemBasePath, 0, strlen($basePath));
+
+            $resource = $this->createResource($filesystemPath);
+
+            if ($resource instanceof FilesystemResource) {
+                $resource->attachTo($this, $path);
+
+                return $resource;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Search for resources by querying their path.
+     *
+     * @param string $query        The glob query.
+     * @param bool   $singleResult Should this method stop after finding a
+     *                             first result, for performances.
+     *
+     * @return ArrayResourceCollection The results of search.
+     */
+    private function search($query, $singleResult = false)
+    {
+        $resources = new ArrayResourceCollection();
+
+        /*
+         * If the query is not a glob, return it directly
+         */
+        if (!Glob::isDynamic($query)) {
+            $resource = $this->resolveResource($query);
+
+            if ($resource instanceof Resource) {
+                $resources->add($resource);
+            }
+
+            return $resources;
+        }
+
+        /*
+         * Otherwise, we need to search for paths matching the glob:
+         *      1.  find the root resource of all search results by
+         *          resolving the glob base path
+         *      2.  find all the children of this root resource and
+         *          try to match their path to the query
+         */
+        $basePath = Glob::getBasePath($query);
+        $baseResource = $this->resolveResource($basePath);
+
+        if ($baseResource !== null) {
+            $children = $this->getChildrenRecursive($baseResource);
+
+            $i = 0;
+            $found = false;
+
+            while ($i < $children->count() && !($found && $singleResult)) {
```

Comment:
I changed it because of ScrunitizerCI: using a `while` avoid a condition and is a bit more clearer than foreach about when it stops. Should i ignore the Scrunitizer notice?